server/authentication.js

```diff
@@ -8,42 +16,196 @@ function enforceHTTPS(request, response, next) {
+
+  // Insert link into database
+  const linkSQL = `INSERT INTO links(email, link, timestampz) VALUES ($1, $2, $3)`;
+  const now = new Date();
+  const linkValues = [email, link, now];
+  return pool.query(linkSQL, linkValues)
+    .then(results => link)
+    .catch(err => {
+      console.log('query returned err: ', err);
+      console.log({ error:err });
+    });
 }
 
+function emailLink(mailgunEnv, email, link) {
+  const linkText = link;
+  const loginlinkFilename = path.join(__dirname,'game/emails/loginlink.html.mustache');
+  const html = renderEmail(loginlinkFilename,{linkText});
+
+  const info = {
+    toEmail: email,
+    fromEmail: 'swipe-right-bot@tsl.mit.edu',
+    subject: 'Swipe Right for CS: Login Link'
+  };
+
+  if (process.env.NODE_ENV !== 'production') {
+    if (process.env.NODE_ENV === 'development') {
+      console.log('No emailing except for in production mode. Go to the following link to move forward.');
+      console.log(link);
+    }
+    return Promise.resolve();
+  }
 
-function loginEndpoint(req, res){
+  return new Promise((resolve, reject) => {
+    sendEmail(mailgunEnv, info, html, (err, mailgunResponse) => {
+      if (err) {
+        console.log("Mailgun request (err):\n  " + JSON.stringify(err, null, 2));
+        return reject(err);
+      }
+      console.log("Mailgun returned:\n  " + JSON.stringify(mailgunResponse, null, 2));
+      return resolve();   
+    });
+  });
+}
+
+
+function loginEndpoint(pool, mailgunEnv, request, response){
   // TODO:
   // 1. Check that the email is in the whitelist in the `researchers` database table
   // 2. Insert a new record to the `links` database table
   // 3. Send an email to the researcher with that link in it (generate email with Mustache template, send it with Mailgun)
   // 4. Return a 200 response
-  res.status(405);
-  res.end();
+  
+  const {email} = request.body;
+
+  isOnWhitelist(pool, email)
+    .then(isNotAuthorized => {
+      if (isNotAuthorized) {
+        return response.status(405).end();
+      } else {
+        const domain = getDomain(request);
+        return createLinkAndEmail(pool, mailgunEnv, email, domain)
+          .then(result => response.status(200).end());
+      }
+    })
+    .catch(err => {
+      console.log('loginEndpoint error: ', err);
+      return response.status(500).end();
+    });
+}
+
+function isOnWhitelist(pool, email){
+  const whitelistSQL = 'SELECT * FROM whitelist WHERE email=$1 ORDER BY id ASC LIMIT 1';
+  const whitelistValues = [email];
+  
+  return pool.query(whitelistSQL, whitelistValues)
+    .then(results =>Promise.resolve(results.rowCount === 0));
+}
+
+function createLinkAndEmail(pool, mailgunEnv, email, domain) {
+  return insertLink(pool,email, domain)
+    .then(link => emailLink(mailgunEnv, email, link));
 }
 
-function emailLinkEndpoint(req, res){
+function emailLinkEndpoint(pool, mailgunEnv, req, res){
+  //This doesn't actually send emails it seems. More like tokenGeneratorEndpoint
```

Comment:
This is good to clarify - if you think it'd help to rename the function then just do that, that'd be great.  If not, update the comment to describe what's happening in a way that would make sense to future folks who will see this for the first time.

Reply:
I think from our last conversation about this, the name for the endpoint makes sense. I've added a comment above the function describing what this function does. I'll update this again when I finished the function!